### ISSUE TITLE
ltemplify.py Not recognising pair_coeffs when names inferred from comments

### ISSUE THREAD

**Issue by AJZein** [jewettaij/moltemplate#25]
When being built, `self.l_data_pair_coeffs` and `self.l_data_pairij_coeffs` (I'm not sure if others do this too) are identified by a name given by using `self.atomtypes_int2name` on their atom types
 e.g. line 1996 `i_str = '@atom:' + self.atomtypes_int2name[i]`

However later in the `postprocess3` function when checking whether to delete a line or not the, name is converted by the `Intify` function (line 2992:3005)

```
atomtype = Intify(split_colon[1])
            if ((not (atomtype in self.needed_atomtypes)) and
                (not ((len(self.atomtype_selection) > 0) and
                      BelongsToSel(atomtype, self.atomtype_selection)))):
                del self.l_data_pair_coeffs[i_line]
```

and the `Intify` function fails to recognise the names of the atoms if they were read from comments, causing all pair_coeffs lines to be deleted. 

Suggestion:

Replace the `Intify ` functions in these sections with the `self.atomtypes_name2int`. Since the names are built originally by `self.atomtypes_int2name` this simply reverts them back to ints. 
Example for `self.l_data_pair_coeffs` (line 2992:3005):

```
atomtype = self.atomtypes_name2int[split_colon[1]]
            if ((not (atomtype in self.needed_atomtypes)) and
                (not ((len(self.atomtype_selection) > 0) and
                      BelongsToSel(atomtype, self.atomtype_selection)))):
                del self.l_data_pair_coeffs[i_line]
```

**Comment by jewettaij** [comment]
I can't work on this today, but I'll definitely take a look at this over the weekend.

Thank you so much for your bug reports and for even wading into the ltemplify.py code.  (This particular program was one of the messiest codes I ever wrote.)  I'll take a look at this and resolve this by Monday.

**Comment by AJZein** [comment]
Happy to help, some very useful utilities you have here.

**Comment by jewettaij** [comment]
Thanks AZ
    Incidentally, I'm not able to reproduce the error you are reporting.  I have attached some files (lammps_input_script.in, lammps_data_file.data) you can use to test ltemplify.py. 

[ltemplate_example_2019-11-10.tar.gz](https://github.com/jewettaij/moltemplate/files/3830029/ltemplate_example_2019-11-10.tar.gz)

Please try running it this way:
```
ltemplify.py lammps_input_script.in lammps_data_file.data > output_ltemplify.lt
```
When I run this the "output_ltemplify.lt" file contains pair_coeffs with @atom type names which match the comments in the input data file:
```
write_once("In Settings") {
  pair_coeff @atom:Cend @atom:Cend 0.066 3.5
  pair_coeff @atom:Cmid @atom:Cmid 0.066 3.5
  pair_coeff @atom:Hydr @atom:Hydr 0.03 2.5
}

Can you send me the input files you are using where you are noticing the problem.  Please also let me know how you are invoking ltemplify.py.  (What arguments are you using?)

Thank you for your hard work with this!

```

**Comment by jewettaij** [comment]
Oh wait.  I see.  From the lines of code you are referring to, it looks like this problem only happens when the Pair Coeffs are in the data file.  (Hmm..  I never test for that case.  Perhaps I should.)  I'm going to bed now, but I'll take another look at this tomorrow.

**Comment by jewettaij** [comment]
Incidentally, I can reproduce the problem now.  No need to send me any files.
And you are right.  There are related problems as well in the bond_coeffs, angle_coeffs, dihedral_coeffs, and improper_coeffs.  Thank you so much for catching this (these) bug(s).

**Comment by jewettaij** [comment]
I approved the PR, but much more extensive changes are necessary.  I also need to write some more CI test scripts to cover all the different ways ltemplify.py can be invoked.  (There are at least 4 different combinations of input types which can confuse ltemplify.py:  Coeffs in the data file or in the input scripts.  Comments vs. no comments.) Trying to get this all done tonight.

Ghastly.  This file grew and grew like a blob from a 1950s monster movie far larger than I suspected possible.  (Duplicated code.  I'd make changes in one place, but not another.)  Oh, the horror.

**Comment by jewettaij** [comment]
Okay.
I fixed most of the most egregious inconsistencies in ltemplify.py in commit a4e27412f7b53d5bbaaea2fc6f1bb09313c4e4d6
It passed the basic Travis CI unit test I wrote last year.  It might be a day or two before I test this code rigorously (and include those tests with my CI unit tests).
I will leave this issue open until I do.

Even so, please let me know later on if you spot more errors.
Thank you so much for the help.
-Andrew

**Comment by AJZein** [comment]
That's great to hear, sorry for not including an example file earlier but glad to hear you worked it out.  If I find anything else I'll let you know. Thanks for working this out so quick.

-Regards,
Zein

**Comment by jewettaij** [comment]
I added some unit tests to make sure ltemplify is printing atom names (and bond names, angle names, etc...) correctly and consistently.  Added tests to make sure the "-datacoeffs" argument is working correctly.  (It wasn't.)  Debugged some problems with the "PairIJ Coeffs" section of the data file.  And I also added tests to make sure ltemplify can understand the "group", "set" and "fix shake" commands.

The code for ltemplify.py is still hideous, but at least it works.
Cheers and thanks again.
-A
